### PR TITLE
correct imu_reading, add getter for imu_frame pos_vel

### PR DESCRIPTION
**Pull request by ShahramKhorshidi**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Correct the simulated imu-reading for linear acceleration. Add a getter for imu-frame position and velocity. 

[//]: # "Description of what you did.  If it is more complex, consider to add a"
[//]: # "'Summary' section."

## Tested in squatting demo in mim_estimation. 

[//]: # "Explain how you tested your changes"

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Comment by MaximilienNaveau** [comment]
@nrotella can you confirm this changes?
I agree with with @ShahramKhorshidi but I want to double check with you.

**Review by MaximilienNaveau — APPROVED** [review]

File: src/bullet_utils/wrapper.py [hunk 1]
```diff
@@ -263,6 +264,28 @@ def get_state(self):
...
+            np.array((3,1)): IMU frame velocity expressed in IMU frame.
+        """ 
+        q, dq = self.get_state()
+        
+        base_inertia_pos, base_inertia_quat = pybullet.getBasePositionAndOrientation(self.robot_id)
+        # Get transform between inertial frame and link frame in base
+        base_stat = pybullet.getDynamicsInfo(self.robot_id, -1)
+        base_inertia_link_pos, base_inertia_link_quat = pybullet.invertTransform(base_stat[3], base_stat[4])
+        base_pos, base_quat = pybullet.multiplyTransforms(base_inertia_pos, base_inertia_quat, base_inertia_link_pos, base_inertia_link_quat)
+        rot_base_to_world = np.array(pybullet.getMatrixFromQuaternion(base_quat)).reshape((3, 3))
```
Comment (nrotella) [inline src/bullet_utils/wrapper.py]: I see that this portion results in the rotation matrix from base to world frame, but why do you have to go through all these steps - do you not get the base quaternion in world frame in the first line? I'm not too familiar with how the frames are represented in pybullet, though. Just trying to understand what's going on.

File: src/bullet_utils/wrapper.py [hunk 2]
```diff
@@ -263,6 +264,28 @@ def get_state(self):
...
+
+        base_angvel = dq[3:6]
```
Comment (nrotella) [inline src/bullet_utils/wrapper.py]: Is this angular velocity in base or world frame? Seems to be in base from how you use it to compute the imu_velocity, but then the imu_position seems to use q[:3] as if that position is in world frame.

File: src/bullet_utils/wrapper.py [hunk 3]
```diff
@@ -215,7 +215,8 @@ def get_base_imu_linacc(self):
         base_linvel, base_angvel = pybullet.getBaseVelocity(self.robot_id)
         
         # Transform the base acceleration to the IMU position, in world frame
-        imu_linacc = self.base_linacc + np.cross(self.base_angacc, self.r_base_to_imu) + np.cross(base_angvel, np.cross(base_angvel, self.r_base_to_imu))
+        imu_linacc = self.base_linacc + np.cross(self.base_angacc, rot_base_to_world @ self.r_base_to_imu) +\
+                     np.cross(base_angvel, np.cross(base_angvel, rot_base_to_world @ self.r_base_to_imu))
 
         return self.rot_base_to_imu.dot(rot_base_to_world.T.dot(imu_linacc + np.array([0.0, 0.0, 9.81]))) + self.base_imu_accel_bias + self.base_imu_accel_thermal
```
Comment (nrotella) [inline src/bullet_utils/wrapper.py:221]: All looks good to me. Does it still work as expected if you change self.rot_base_to_imu to something other than the identity, in the case that the axes of the IMU and base frames are not aligned?

**Review by nrotella — APPROVED** [review]
Looks good to me overall, thanks for adding me. I just had a few questions (more for my understanding), but I assume the code functions as expected and can be merged based on the plots Shahram has shared recently.

File: src/bullet_utils/wrapper.py (hunk 1 above)
Comment (ShahramKhorshidi) [reply src/bullet_utils/wrapper.py]: Thanks for the comments. Yes that's right, I can get base quaternion from the first line. I modified the code, now the base velocity is in world. 
The returned value of imu_position is expressed in world and imu_velocity is in IMU frame.

File: src/bullet_utils/wrapper.py (hunk 3 above)
Comment (ShahramKhorshidi) [reply src/bullet_utils/wrapper.py:221]: Yes, I also ran the ekf in the case which imu is rotated w.r.t base, and the results are correct.

**Comment by jviereck** [comment]
Thanks for working on this @ShahramKhorshidi !

**Pull request merged**